gradle-guide-error-prone/src/main/java/com/palantir/gradle/guide/errorprone/ConfigurationAvoidanceRegistration.java

```diff
@@ -0,0 +1,136 @@
+                + "and performing unnecessary work which will slow down the build.")
+public final class ConfigurationAvoidanceRegistration extends GradleGuideBugChecker
+        implements BugChecker.MethodInvocationTreeMatcher {
+    private static final Matcher<ExpressionTree> MATCHER = Matchers.instanceMethod()
+            .onDescendantOfAny("org.gradle.api.NamedDomainObjectContainer")
+            .namedAnyOf("create");
+
+    private static final Matcher<MethodInvocationTree> FIRST_ARGUMENT_IS_MAP =
+            Matchers.argument(0, Matchers.isSubtypeOf(Map.class));
+
+    private static final Matcher<MethodInvocationTree> SECOND_ARGUMENT_IS_GROOVY_CLOSURE =
+            Matchers.argument(1, Matchers.isSubtypeOf("groovy.lang.Closure"));
+
+    private static final Matcher<MethodInvocationTree> NO_DIRECT_REGISTER_EQUIVALENT =
+            Matchers.anyOf(FIRST_ARGUMENT_IS_MAP, SECOND_ARGUMENT_IS_GROOVY_CLOSURE);
+
+    @Override
+    public Description matchMethodInvocation(MethodInvocationTree tree, VisitorState state) {
+        if (!MATCHER.matches(tree, state)) {
+            return Description.NO_MATCH;
+        }
+
+        Description.Builder description = buildDescription(tree);
+        SuggestedFixBuilder fix = new SuggestedFixBuilder();
+
+        if (!bestEffortModeEnabled(state)) {
+            return description.build();
+        }
+
+        // If the first argument is a map, or second is a Closure, there isn't an equivalent
+        // `.register` method to move to (from Java code at least)
+        if (NO_DIRECT_REGISTER_EQUIVALENT.matches(tree, state)) {
+            return description.build();
+        }
+
+        TreePath parentPath = state.getPath().getParentPath();
+        if (parentPath.getLeaf() instanceof VariableTree variableTree) {
+            replaceVariableDeclartionTypeWithProvider(state, fix, variableTree);
+            replaceVariableUsagesWithTaskProviderGet(fix, parentPath);
+        }
+
+        fix.replace(
+                tree.getMethodSelect(),
+                state.getSourceForNode(ASTHelpers.getReceiver(tree.getMethodSelect())) + ".register");
+
+        if (parentPath.getLeaf() instanceof ExpressionTree || parentPath.getLeaf() instanceof ReturnTree) {
+            fix.postfixWith(tree, ".get()");
+        }
+
+        return description.addFix(fix.build()).build();
+    }
+
+    private static void replaceVariableDeclartionTypeWithProvider(
+            VisitorState state, SuggestedFix.Builder fixBuilder, VariableTree variableTree) {
+
+        boolean isTask = state.getTypes()
+                .isSubtype(ASTHelpers.getType(variableTree.getType()), state.getTypeFromString("org.gradle.api.Task"));
+
+        if (isTask) {
+            fixBuilder.addImport("org.gradle.api.tasks.TaskProvider");
+            fixBuilder.prefixWith(variableTree.getType(), "TaskProvider<");
+        } else {
+            fixBuilder.addImport("org.gradle.api.provider.Provider");
```

Comment:
do you not want this to be a `NamedDomainObjectProvider` ?

Reply:
Could be, I actually prefer using `Provider` unless you need the `.getName()` as it isn't so horribly verbose. That said, for best effort mode it'd probably be better as `NamedDomainObjectProvider` _if_ we're able to automatically make it use `.getName()` somehow.

Reply:
Made it a NDOP, realised we probably need it for `.configure` as well.